aws/amazon-freertos
# August 28, Added Router Advertisement/SLAAC

Description
-----------
Until now, IPv6 end-points had to be configured statically.
This PR makes it possible to solicit for a Router Advertisement, which may provide the following parameters:

- Network prefix
- Prefix length
- Address of a IPv6 gateway

When these parameters are known, the device will choose a **random IP-address that uses the prefix and prefix-length**.

It will be tested if the IP-address already exists by issuing a few neighbour solicitations, and wait for replies.
When no other device replies, the address will be taken and the end-point will go UP. When there are replies, a new one will be chosen, and the procedure is repeated.

These are the default configuration macro's that control the entire process:

~~~c
    /* Ask for a Router Advertisement at most 3 times. */
    #define ipconfigRA_SEARCH_COUNT             ( 3 )
    /* Each time waiting at most 10 seconds for an answer. */
    #define ipconfigRA_SEARCH_TIME_OUT_MSEC     ( 10000u )

    /* Test the self-chose IP-address 3 times. */
    #define ipconfigRA_IP_TEST_COUNT            ( 3 )
    /* Each time waiting at most 1.5 seconds for a reply. */
    #define ipconfigRA_IP_TEST_TIME_OUT_MSEC    ( 1500u )
~~~

All sources were PC-LINT-ed, and tested in several ways, both with IPv4 and IPv6:

- Ping to the Internet
- Connect to a web-site
- Receive pings from the Internet [using this tool](https://www.ipaddressguide.com/ping6)
- Connect from a laptop to the device


Checklist:
----------


- [ x ] I have tested my changes. No regression in existing tests.
- [ x ] My code is Linted.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
